refactor postgres modules, sslmode=prefer

tedder opened the pull request:
- ssl_mode is new, the default in Ansible modules was 'disable'. This doesn't match the Postgres default of 'prefer'. Change per irc discussion with abadger.
- refactor to use a common postgresql base util. Tried to keep any aliases.
- refactor to use a common docs fragment.
- fix some mismatched documentation (e.g., login_user didn't have a default in docs, but it is 'postgres' in code). more docfixes needed.
- add no_log to password.

Tested with three databases, so I could test the ssl modes in a bunch of various configurations. I'll show my test play in the PR.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```

```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
```

```

tedder committed:
refactor postgres modules, sslmode=prefer

- ssl_mode is new, the default in Ansible modules was 'disable'. This doesn't match the Postgres default of 'prefer'. Change per irc discussion with abadger.
- refactor to use a common postgresql base util. Tried to keep any aliases.
- refactor to use a common docs fragment.
- fix some mismatched documentation (e.g., login_user didn't have a default in docs, but it is 'postgres' in code). more docfixes needed.
- add no_log to password.

Tested with three databases, so I could test the ssl modes in a bunch of various configurations. I'll show my test play in the PR.

tedder commented:
Here's my scrubbed play:

```
# simple local docker container for nonssl testing:
# docker run -e POSTGRES_PASSWORD=secretpass -e POSTGRES_USER=myuser -p 5432:5432 -it postgres:9.6-alpine
#
# psql connection reminder:
# PGPASSWORD=secretpass psql -hlocalhost -U myuser postgres

- name: extensive pg_db tests. others will be simpler.
  hosts: localhost
  connection: local
  gather_facts: false
  tasks:
  # local db is non-ssl-only.
  # tests that should (and do) succeed
  - postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present
  - postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=disable
  - postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=allow
  - postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=prefer

  # expected to fail, and does
  #- postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=require
  # don't have infra handy to test this (need root cert)
  #- postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=verify-ca
  #- postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=verify-full

  # clean up.
  - postgresql_db: name=testdb login_password=secretpass login_user=myuser login_host=localhost state=absent

  # ssl-only, obv. dbname that already exists: sslonly666
  - postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=require
  - postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=allow
  - postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=prefer

  # expected to fail, and does
  #- postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=disable
  # don't have infra handy to test this (need root cert)
  #- postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=verify-ca
  #- postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=verify-full
  # clean up
  - postgresql_db: name=testdb login_password=sslpass1 login_user=ssluser login_host=test-sslonly-db ssl_mode=require

  # ssl-optional, obv. dbname that already exists: ssloptional667
  # tested modes that fail (as expected):
  # tested modes that succeed (as expected):
  - postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=disable
  - postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=allow
  - postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=prefer
  - postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=require

  # don't have infra handy to test this (need root cert)
  #- postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=verify-ca
  #- postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=verify-full

  # clean up
  - postgresql_db: name=testdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=require


- name: pg ext tests
  hosts: localhost
  connection: local
  gather_facts: false
  tags: ext
  tasks:
  - name: create DB so the following will work
    postgresql_db: name=extdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db

  - postgresql_ext: name=ltree db=extdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db
  - postgresql_ext: name=ltree db=extdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db ssl_mode=require
  # clean up
  - postgresql_ext: name=ltree db=extdb login_password=sslpass2 login_user=optionalssluser login_host=test-ssloptional-db state=absent


- name: pg lang tests
  hosts: localhost
  connection: local
  gather_facts: false
  tags: lang
  tasks:
  - name: create DB so the following will work
    postgresql_db: name=langdb login_password=secretpass login_user=myuser login_host=localhost state=present

  # we are adding a lang that should alreaady be present, so no cleanup afterwards.
  - postgresql_lang: lang=plpgsql db=langdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=prefer
  # fails locally as expected
  #- postgresql_lang: lang=plpgsql db=langdb login_password=secretpass login_user=myuser login_host=localhost state=present ssl_mode=require



- name: pg schema tests
  hosts: localhost
  connection: local
  gather_facts: false
  tags: schema
  tasks:
  - name: create DB so the following will work
    postgresql_db: name=schemadb login_password=secretpass login_user=myuser login_host=localhost state=present

  - postgresql_schema: name=myschema database=schemadb login_password=secretpass login_user=myuser login_host=localhost state=present
  - postgresql_schema: name=myschema database=schemadb login_password=secretpass login_user=myuser login_host=localhost state=absent

- name: pg user tests
  hosts: localhost
  connection: local
  gather_facts: false
  tags: user
  tasks:
  - name: create DB so the following will work
    postgresql_db: name=userdb login_password=secretpass login_user=myuser login_host=localhost state=present

  - postgresql_user: db=userdb name=user42 password=secretpass42 fail_on_user=true login_password=secretpass login_user=myuser login_host=localhost state=present
  - postgresql_user: db=userdb name=user42                       fail_on_user=true login_password=secretpass login_user=myuser login_host=localhost state=absent

- name: pg privs tests. needs pg_user to be working.
  hosts: localhost
  connection: local
  gather_facts: false
  tags: privs
  tasks:
  - name: create DB so the following will work
    postgresql_db: name=privdb login_password=secretpass login_user=myuser login_host=localhost state=present
  - name: create user so the following will work
    postgresql_user: db=privdb name=user2 password=secretpass42 fail_on_user=true login_password=secretpass login_user=myuser login_host=localhost state=present

  - postgresql_privs: db=privdb privs=ALL type=database roles=user2 login_password=secretpass login_user=myuser login_host=localhost state=present


```

tedder commented:
cc @abadger, waiting for shipping before I bug you on IRC.

tedder committed:
rm commented-out bits from stealing code

tedder committed:
rm postgres from legacy

tedder committed:
pep8 fixes: line length, bad indents, comparing to false

tedder committed:
pep8 whitespace and line length fixes

tedder committed:
use module.params to check for sslrootcert

tedder committed:
aha. localhost should not be defaulted. needs to be blank, that lets psycopg default to UNIX socket.

tedder committed:
use NotSupportedError from pgutils only.

tedder committed:
aha, need to ensure we are on the correct db

tedder committed:
rm duplicated param

tedder commented on lib/ansible/module_utils/postgres.py:
```diff
@@ -0,0 +1,119 @@
+
+    return kw
+
+def postgres_conn(module, database=None, kw=None, enable_autocommit=False, **params):
```
sample error that probably comes from here:
```unable to connect to database: 'sslrootcert'```
 https://app.shippable.com/runs/589aa0bd6ee43c0f00b543cd/9/console

tedder committed:
rm params, set reasonable default for kw

nerzhul commented:
interesting cleanup

mattclay commented:
CI failure due to:

```
2017-02-08 05:05:03 TASK [postgresql : Create DB] **************************************************
2017-02-08 05:05:03 task path: /root/ansible/test/integration/targets/postgresql/tasks/main.yml:4
2017-02-08 05:05:03 Using module file /root/ansible/lib/ansible/modules/database/postgresql/postgresql_db.py
2017-02-08 05:05:03 <127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
2017-02-08 05:05:03 <127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /tmp/ansible-tmp-1486530303.97-197632969627292 `" && echo ansible-tmp-1486530303.97-197632969627292="` echo /tmp/ansible-tmp-1486530303.97-197632969627292 `" ) && sleep 0'
2017-02-08 05:05:03 <127.0.0.1> PUT /tmp/tmpBwDsQb TO /tmp/ansible-tmp-1486530303.97-197632969627292/postgresql_db.py
2017-02-08 05:05:03 <127.0.0.1> EXEC /bin/sh -c 'setfacl -m u:postgres:r-x /tmp/ansible-tmp-1486530303.97-197632969627292/ /tmp/ansible-tmp-1486530303.97-197632969627292/postgresql_db.py && sleep 0'
2017-02-08 05:05:03 <127.0.0.1> EXEC /bin/sh -c 'sudo -H -S -n -u postgres /bin/sh -c '"'"'echo BECOME-SUCCESS-kezoljxlarsciyqsgbgosnlghswigigp; /tmp/ansible-test-coverage-ixhijR/coverage/runner2.6 /tmp/ansible-tmp-1486530303.97-197632969627292/postgresql_db.py'"'"' && sleep 0'
2017-02-08 05:05:04 <127.0.0.1> EXEC /bin/sh -c 'rm -f -r /tmp/ansible-tmp-1486530303.97-197632969627292/ > /dev/null 2>&1 && sleep 0'
2017-02-08 05:05:04 fatal: [testhost]: FAILED! => {
2017-02-08 05:05:04     "changed": false, 
2017-02-08 05:05:04     "failed": true, 
2017-02-08 05:05:04     "invocation": {
2017-02-08 05:05:04         "module_args": {
2017-02-08 05:05:04             "db": "ansible_db", 
2017-02-08 05:05:04             "encoding": "", 
2017-02-08 05:05:04             "lc_collate": "", 
2017-02-08 05:05:04             "lc_ctype": "", 
2017-02-08 05:05:04             "login_host": "", 
2017-02-08 05:05:04             "login_password": "", 
2017-02-08 05:05:04             "login_unix_socket": "", 
2017-02-08 05:05:04             "login_user": "postgres", 
2017-02-08 05:05:04             "name": "ansible_db", 
2017-02-08 05:05:04             "owner": "", 
2017-02-08 05:05:04             "port": 5432, 
2017-02-08 05:05:04             "ssl_mode": "prefer", 
2017-02-08 05:05:04             "ssl_rootcert": null, 
2017-02-08 05:05:04             "state": "present", 
2017-02-08 05:05:04             "template": ""
2017-02-08 05:05:04         }
2017-02-08 05:05:04     }, 
2017-02-08 05:05:04     "msg": "unable to connect to database: 'sslrootcert'"
2017-02-08 05:05:04 }
```

Logs here: https://app.shippable.com/runs/589aa5ed5512b90f00b73300/9/console

tedder commented:
Yeah, @mattclay, I'm guessing the "kw" dict is getting mangled. You can see I've made an embarrassing number of attempts to fix it :/ I can't even repro it.

alikins commented on lib/ansible/module_utils/postgres.py:
```diff
@@ -0,0 +1,119 @@
+
+    return kw
+
+def postgres_conn(module, database=None, kw={}, enable_autocommit=False):
```
lib/ansible/module_utils/postgres.py|74 col 1 warning| - mutable default arg of type Dict [M511] [python/flake8]

I would avoid using mutable default args (ie, kw={}) 

try

``` python
def postgres_conn(module, database=None, kw=None, enable_autocommit=False):
    kw = kw or {}
```

alikins commented on lib/ansible/module_utils/postgres.py:
```diff
@@ -0,0 +1,119 @@
+            module.fail_json(msg='Postgresql server must be at least version 8.4 to support sslrootcert')
+        module.fail_json(msg="unable to connect to database: %s" % e)
+
+    except Exception:
```
For test/debug, I would suggest adding a 'raise' here. Otherwise any exception is going to get obscured.

(For ex, I suspect the 'sslrootcert' error is actually a KeyError that makes it back to this except clause)

alikins commented on lib/ansible/module_utils/postgres.py:
```diff
@@ -0,0 +1,119 @@
+def postgres_conn(module, database=None, kw={}, enable_autocommit=False):
+
+    try:
+        if psycopg2.__version__ < '2.4.3' and module.params['sslrootcert'] is not None:
```
Guessing this should be a module.params.get('sslrootcert', None) otherwise it will be a KeyError (see comment below)

tedder committed:
don't fail_json directly, other CR fixes

I'll need to add try/except back to the pg modules now, but I want to get this committed as a response and to let Shippable do its thing to see if this fixes the weird sslrootcert failure.

alikins commented on lib/ansible/module_utils/postgres.py:
```diff
@@ -65,15 +65,17 @@ def params_to_kwmap(module):
     return kw
 
-def postgres_conn(module, database=None, kw={}, enable_autocommit=False):
+def postgres_conn(module, database=None, kw=kw, enable_autocommit=False):
```
``` python
def postgres_conn(module, database=None, kw=None, enable_autocommit=False):
```

tedder committed:
reverse the {} and None params. oops.

alikins commented:
I'd suggest starting with some unit tests for module_utils/postgresql.py
For ex,

test/units/module_utils/test_postgresql.py
``` python

import json

from ansible.compat.tests import unittest

from ansible.module_utils import postgres
from ansible.module_utils.basic import AnsibleModule
from units.mock.procenv import swap_stdin_and_argv


import pprint


class TestPostgres(unittest.TestCase):
    def test(self):
        params = {'foo': 'blippy'}
        # module weirdness and testing shenagigans
        args = json.dumps(dict(ANSIBLE_MODULE_ARGS=params))
        self.stdin_swap = swap_stdin_and_argv(stdin_data=args)
        self.stdin_swap.__enter__()

        a_module = AnsibleModule(argument_spec={'foo': {'required': True, 'aliases': ['name']}})
        res = postgres.params_to_kwmap(a_module)
        pprint.pprint(res)

        # unittest doesn't have a clean place to use a context manager, so we have to enter/exit manually
        self.stdin_swap.__exit__(None, None, None)
```

tedder committed:
safe-get on dict entry

tedder committed:
add basic unit test, thx alikins

tedder committed:
use test_basic as a template for the magic swapfoo

tedder committed:
add missing import

tedder commented:
Closing, I've replaced this with #21435 and #21498. I'll continue refactoring them a little more cautiously.
